src/main/java/training/chessington/model/pieces/AbstractPiece.java

```diff
@@ -26,4 +36,58 @@ public PlayerColour getColour() {
     public String toString() {
         return colour.toString() + " " + type.toString();
     }
+
+    protected List<Move> getDiagonalMoves(Coordinates from) {
+        List<Function<Coordinates, Coordinates>> directionVectors = new ArrayList<>();
+        directionVectors.add((initialPosition) -> initialPosition.plus(1, 1));
+        directionVectors.add((initialPosition) -> initialPosition.plus(1, -1));
+        directionVectors.add((initialPosition) -> initialPosition.plus(-1, -1));
+        directionVectors.add((initialPosition) -> initialPosition.plus(-1, 1));
+        return this.getLinearMovesManyDirections(from, directionVectors);
+    }
+
+    protected List<Move> getHorizontalMoves(Coordinates from) {
+        List<Function<Coordinates, Coordinates>> directionVectors = new ArrayList<>();
+        directionVectors.add((initialPosition) -> initialPosition.plus(0, 1));
+        directionVectors.add((initialPosition) -> initialPosition.plus(0, -1));
+        return this.getLinearMovesManyDirections(from, directionVectors);
+    }
+
+    protected List<Move> getVerticalMoves(Coordinates from) {
+        List<Function<Coordinates, Coordinates>> directionVectors = new ArrayList<>();
+        directionVectors.add((initialPosition) -> initialPosition.plus(1, 0));
+        directionVectors.add((initialPosition) -> initialPosition.plus(-1, 0));
+        return this.getLinearMovesManyDirections(from, directionVectors);
+    }
+
+    private List<Move> getLinearMovesManyDirections(Coordinates initial, Iterable<Function<Coordinates, Coordinates>> directionVectors) {
+        List<Move> linearMoves = new ArrayList<>();
+        for (Function<Coordinates, Coordinates> directionVector: directionVectors) {
+            linearMoves.addAll(this.getLinearMovesOneDirection(initial, directionVector));
+        }
+        return linearMoves;
+    }
+
+    private List<Move> getLinearMovesOneDirection(Coordinates initialPosition, Function<Coordinates, Coordinates> directionVector) {
+        List<Move> linearMoves = new ArrayList<>();
+        Coordinates nextPosition = directionVector.apply(initialPosition);
+        while (board.contains(nextPosition)) {
+            linearMoves.add(new Move(initialPosition, nextPosition));
+            nextPosition = directionVector.apply(nextPosition);
+        }
+        return linearMoves;
+    }
+
+    protected List<Move> removeInvalidMoves(List<Move> moveList) {

```

Comment:
If you intend for this method to be called by every subclass, then there might be a better approach. Currently, you have to remember to call it inside the `getAllowedMoves` of each subclass, and you might forget.

How about defining a new abstract method called, say, `getAllowedMovesBeforeValidityChecks`, and have each subclass implement that method. Then make the original `getAllowedMoves` method no longer abstract and define its implementation in terms of the other method.

That's the right thing to do in theory, anyway (assuming I've understood the purpose of this method correctly). In practice, I'm not sure about how useful it really is given most overrides of it appear trivial (`return true`).